File: specs/pkg-go/trustless-pathing/unixfs_20m_variety.go
package trustless_pathing

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/ipfs/go-cid"
	carstorage "github.com/ipld/go-car/v2/storage"
	"github.com/ipld/go-ipld-prime/storage"
	"github.com/warpfork/go-testmark"

	_ "github.com/ipld/go-codec-dagpb"
	_ "github.com/ipld/go-ipld-prime/codec/raw"
)

const file = "/../../transport/trustless-pathing/fixtures/unixfs_20m_variety."

func pathToFixture(typ string) string {
	_, thisFile, _, _ := runtime.Caller(0)
	pathTo := filepath.Dir(thisFile) + file + typ
	pathTo = filepath.Clean(pathTo)
	// convert wd to absolute path, normalizing to / separators
	pathTo = strings.ReplaceAll(pathTo, "\\", "/")
	return pathTo
}

func Unixfs20mVarietyCARPath() string {
	return pathToFixture("car")
}

func Unixfs20mVarietyReadableStorage() (storage.ReadableStorage, io.Closer, error) {
	file := pathToFixture("car")
	carFile, err := os.Open(file)
	if err != nil {
		return nil, nil, err
	}
	reader, err := carstorage.OpenReadable(carFile)
	if err != nil {
		carFile.Close()
		return nil, nil, err
	}
	return reader, carFile, nil
}

func Unixfs20mVarietyCases() ([]TestCase, cid.Cid, error) {
	file := pathToFixture("md")
	doc, err := testmark.ReadFile(file)
	if err != nil {
		return nil, cid.Undef, fmt.Errorf("failed to read testcases: %w", err)
	}
	doc.BuildDirIndex()
	root, err := cid.Parse(dstr(doc.DirEnt, "root"))
	if err != nil {
		return nil, cid.Undef, err
	}
	testCases := make([]TestCase, 0)
	for _, test := range doc.DirEnt.Children["test"].ChildrenList {
		for _, scope := range test.ChildrenList {
			tc, err := ParseCase(test.Name+"/"+scope.Name, dstr(scope, "query"), dstr(scope, "execution"))
			if err != nil {
				return nil, cid.Undef, fmt.Errorf("failed to parse test case %s: %w", test.Name+"/"+scope.Name, err)
			}
			testCases = append(testCases, tc)
		}
	}
	return testCases, root, nil
}

func dstr(dir *testmark.DirEnt, ch string) string {
	return string(dir.Children[ch].Hunk.Body)
}

File: specs/pkg-js/trustless-pathing/unixfs_20m_variety.js
import { readFileSync } from 'fs'
import { parseCase } from './testcase.js'
import { parse, index } from 'testmark.js'

const file = '../../transport/trustless-pathing/fixtures/unixfs_20m_variety.'

function pathToFixture (typ) {
  return new URL(file + typ, import.meta.url)
}

export function unixfs20mVarietyCar () {
  return pathToFixture('car')
}

export function unixfs20mVarietyCases () {
  const file = pathToFixture('md')
  const contents = readFileSync(file, 'utf-8')
  const dirEnt = index(parse(contents))
  const testCases = []
  for (const test of dirEnt.children.get('test').childrenList) {
    for (const scope of test.childrenList) {
      testCases.push(parseCase(`${test.name}/${scope.name}`, dstr(scope, 'query'), dstr(scope, 'execution')))
    }
  }
  return testCases
}

function dstr (dir, ch) {
  return dir.children.get(ch).hunk.body
}
